### ISSUE TITLE
Feature: Set Image With Code?

### ISSUE THREAD

**Issue by quicklywilliam** [onevcat/Kingfisher#917]
### Issue Description

In our app, we often generate images from code. Operationally, generating images from code closely resembles downloading them from the server (async, background thread, caching etc). It would be nice to be able to leverage Kingfisher's operational and caching mechanisms for this use case. Note that this is different from a Processor because it does not necessarily start from a base image but rather from custom app data or state.

#### What
I'm imagining something like the following:

```

imageView.kf.setImage(with: { (computedImageData) in
  UIGraphicsBeginImageContextWithOptions(computedImageData.imageSize, computedImageData.opaque , 0.0)
    // do things, render the image
  let image = UIGraphicsGetImageFromCurrentImageContext()
  UIGraphicsEndImageContext()
  return image
})

```

Here computedImageData would be an instance of ComputedImageData, a simple class that could encapsulate both standard data (like image size and opaqueness) and custom data via subclassing. Something like this:
```
class ComputedImageData {
  public var imageSize: CGSize = CGSize.zero
  public var opaque: Bool = false
}
```

#### Discussion

We'd be happy to work on a pull request for this if there is sufficient interest. For now, I wanted to see if this has been proposed before and whether folks thought this is within the scope of what Kingfisher should do.

**Comment by onevcat** [comment]
Hi, @quicklywilliam 

Thanks for the suggestion. This seems to be under a sub-domain of implement #722 , however, currently Kingfisher heavily depends on the URL to download the image, so under current architecture implementing an arbitrary customizable resource is not so trivial. (I plan to find time to refactoring the whole project and release a major update for that.)

For your particular case, what you want is the cache part of Kingfisher (which is managed by `ImageCache`). So once you have a way to give the image an identifier, you could add an extension to `UIImageView` easily. Maybe something like this:

```swift
class ComputedImageData {
    public var imageSize: CGSize = CGSize.zero
    public var opaque: Bool = false
    public var identifier: String = ""
}

extension Kingfisher where Base: ImageView {
    func setImage(with resource: ComputedImageData,
                         placeholder: Placeholder? = nil,
                         options: KingfisherOptionsInfo? = nil,
                         progressBlock: DownloadProgressBlock? = nil,
                         completionHandler: CompletionHandler? = nil)
    {
        let options = KingfisherManager.shared.defaultOptions + (options ?? [])
        options.targetCache.retrieveImage(forKey: resource.identifier, options: options) { (image, type) in
            if let image = image {
                self.base.image = image
                completionHandler?(image, nil, type, nil)
            } else {
                UIGraphicsBeginImageContextWithOptions(resource.imageSize, resource.opaque , 0.0)
                // do things, render the image
                let image = UIGraphicsGetImageFromCurrentImageContext()!
                options.targetCache.store(image, forKey: resource.identifier)
                self.base.image = image
                completionHandler?(image, nil, .none, nil)
            }
        }
    }
}
```

**Comment by onevcat** [comment]
It should be added from version 5 (not released yet, but you can try it from branch 5.0 or some beta release later). Use the `ImageDataProvider` would do the trick.
